### ISSUE TITLE
Idea: Thread pool tasks have a "home" worker.

### ISSUE THREAD

**Issue by carllerche** [tokio-rs/tokio#424]
Currently, the runtime uses a thread pool and an extra dedicated thread to run a reactor instance. The reactor is used to drive all sockets (and other Mio backed I/O resources). This layout means that for very inbound Mio event, there will be cross thread synchronization incurred when the reactor notifies the associated task that it should be polled.

The reason for the single reactor instance on a dedicated thread is that once a socket is registered with a Mio instance, it stays registered for its lifetime.

The idea proposed in this issue is to have one reactor instance per thread. When a socket is registered with a reactor, the one for the current thread is used. The assumption is that sockets are not often migrated across tasks. In the event that a socket is migrated to a new task that is polled on another thread, then the the scenario is no worse than it is today.

## A "home" worker

In order for sockets to stay on the same thread as their reactor, tasks would be assigned a "home" worker. When a task is scheduled, it is always scheduled to execute on its home worker. While it is sitting in that worker's work queue, it might be stolen by another worker. However, eventually it would find its way back to home.

The home worker would be assigned to a task the first time it is polled. If a task is stolen before it is polled the first time, the thief becomes the home.

## Benefits

As described, the main benefit is to reduce the amount of synchronization between the reactor and tasks that use resources driven by a reactor. However, there is an additional benefit: it enables solving #428. Because a future is now assigned to a worker, that worker is able to track all futures that it owns, enabling the worker to drop incomplete futures when it shuts down.

## Open questions

How would this change impact the thread pool's ability to balance its load across workers? Right now, whenever a future is notified, it is scheduled on a random worker. This issue proposes changing that. It is possible that a) this is not an issue in practice or b) additional stealing heuristics could be developed to mitigate. For example, workers could attempt to steal from other workers that are under heavier load even when they are not idle.

Another question is how would this work with back up threads? There should be a reactor per worker thread and not per back up thread. It may not be a problem, but it is worth thinking about for a moment.

**Comment by fpgaminer** [comment]
FYI, I believe I ran into this issue with a very simple program (my first Tokio program actually...).  Context here:  https://users.rust-lang.org/t/getting-tokio-to-match-actix-web-performance/18659/10

TL;DR:  Simple HTTP server gets poor performance in artificial benchmarks with default Tokio setup.  Much better performance by spawning a CurrentThread runtime per core so that sockets live on the thread they're accepted on.

**Comment by jonhoo** [comment]
@fpgaminer I recently ran into something similar and built [`tokio-io-pool`](https://docs.rs/tokio-io-pool/), which provides a convenient way to run futures on an I/O focused thread pool (i.e., one where every thread has its own I/O reactor and runtime).

**Comment by naftulikay** [comment]
From my (very limited) understanding of things, it would seem that the highest throughput would be achieved by successfully implementing cross-thread work stealing for both CPU and I/O bound tasks. Obviously, the goal of work-stealing is to keep tasks local to a specific thread as much as possible, but steal tasks across threads when threads get backlogged.

Synchronization overhead is mentioned above... I'd imagine this wouldn't be a lot of overhead, provided that we're not using `pthread_mutex_lock` and preferring atomic values similar to [parking_lot](https://github.com/amanieu/parking_lot). However, would this mean a lot of explicit `Arc`s being created?

My limited understanding seems to tell me that the following chain of improvements can be followed to get the highest theoretical concurrency throughput:

 1. Rather than a single thread, multiple threads.
 1. Rather than unbounded threads, a fixed number of threads relative to the workload/CPU count.
 1. Rather than being limited by OS threads, implementing an M:N threading model with _m_ OS threads and _n_ threads in userspace, similar to Go's coroutines.
 1. Rather than blocking I/O in threads and coroutines, using libaio/epoll/etc. to notify a single thread when an I/O bound task is ready to be reacted to.
 1. Rather than doing event polling on a single I/O thread, having multiple reactors that are isolated and pinned to their host thread.
 1. Rather than pinning tasks to the thread they're created in, allow work-stealing between threads to keep the load distributed as fairly as is reasonably possible.

It seems that with [`tokio-io-pool`](https://docs.rs/tokio-io-pool/), we have arrived at step 5 :ok_hand:

Are there any other improvements that can be made beyond removing the requirement of pinning tasks/coroutines/IO to the thread they were created on? Obviously, this is contingent on having a good work-stealing algorithm, but maybe I'm missing something.

It's exciting to see how far Tokio has come :+1:

**Comment by jonhoo** [comment]
I think `tokio::runtime::Runtime` provides all of 1-4 already, but yes, I think the next step would be to add work-stealing of I/O resources. It's not immediately obvious to me whether the "right" thing to do is to start from a model like the current `tokio::Runtime` and then split up the I/O processing, or start with the one-`Runtime`-per-thread and then add work-stealing to that. They both have challenges, though I think ultimately they both face the challenge of needing to keep track of I/O tasks, and being able to de-schedule them from the underlying reactor and then re-registering with a different reactor elsewhere. I think @carllerche has had some thoughts around this; perhaps he could share those here?

**Comment by naftulikay** [comment]
@jonhoo so in the docs, it says that `tokio-io-pool` is heavily tuned for I/O heavy tasks. While it's certainly better than the default pool, will CPU-bound tasks perform generally as well as they do in the default pool implementation?

**Comment by Bert-Proesmans** [comment]
I tried to discuss something similar on the user forum, but my lack of understanding kind off silenced the thread.  
What I learned was that the response latency becomes unstable as a trade-off. There is also no guarantee (for now) that task stealing would be stable; how would you avoid a task from constantly being stolen by another worker? An even bigger problem when workers can steal without being idle as precondition.

https://users.rust-lang.org/t/combining-work-stealing-and-so-reuseport/18717

**Comment by jonhoo** [comment]
In general, workers only steal when they have no other work to do. @carllerche can probably give the exact conditions, but you shouldn't generally see tasks bounce back and forth between workers. In `tokio-io-pool` (#486), there is no stealing, and the work is basically distributed the same as it sounds like it is in actix-web. The thread that is first given a stream when it is accepted continues processing it forever.

**Comment by jonhoo** [comment]
To clarify: Tokio today has work-stealing for all *tasks*, but all I/O resources (i.e., syscalls) are driven by a single reactor on one thread. This effectively means that you're bottlenecked by how fast a single thread can make syscalls, which is somewhere around 2M/s on a recent system. And effectively, it's lower than that because of additional synchronization to the reactor and additional calls like `epoll` that the reactor also has to do.

**Comment by carllerche** [comment]
@jonhoo Not quite. The only syscall that happens on the reactor thread is `epoll_wait`. All other syscalls happen on the threadpool threads. So, if you use a socket on a pool thread, read / write calls will happen there. If the task gets stolen to another thread, those syscalls will move as well.

In practice, the reactor thread is never the bottleneck. The main issue is the synchronization required between the reactor thread and the threadpool threads.

**Comment by carllerche** [comment]
@fpgaminer I would also warn of the danger of micro benchmarks. Synchronization required by work-stealing will almost always result in slower performance in micro benchmarks. This is because the work load tends to be very uniform and tiny. However, there is a reason that a work-stealing approach is used by virtually all schedulers. Once in the context of a real application, things aren't as uniform as the micro benchmark.

**Comment by jonhoo** [comment]
@carllerche Oh interesting, I'd missed that. I guess that makes sense now that I think about it — the tasks call `AsyncRead::poll_read` and `AsyncWrite::poll_write`, which is what does the syscalls, and the tasks are what move around. Makes me wonder why I see such large speedups with tokio-io-pool... One hypothesis would be that the reactor thread has to spend a bunch of time notifying any awoken tasks (i.e., synchronization), which is latency it adds to when it next calls `epoll_wait`. This gets particularly bad if you have many connections and each connection is has only a single request-response in flight at a time. Lots of tasks are woken up, they all do a single read+write, and then all await notification again. So now they need to wait for the reactor to finish doing all the notifies, and then for it to do `epoll_wait`.

This suggests that we could significantly improve performance (at least in my project) by having fewer connections and multiplexing operations onto them...

**Comment by lnicola** [comment]
The comments here and #486, and what @seanmonstar wrote seem to assume that doing this must mean that work-stealing can no longer be used. Are they really exclusive?

Are there any issues with having a reactor per thread, queuing the tasks to itself, while allowing workers to steal from the others when idle?

Also, I vaguely recall reading (or dreaming) that .NET's thread pool workers steal half of the items instead of a single one.

**Comment by seanmonstar** [comment]
I didn't try to imply that having a home worker means a thread pool cannot do work-stealing. I was saying that the implementation of tokio-io-pool *at that time* meant that it doesn't do work-stealing, that's all.

**Comment by naftulikay** [comment]
Also, work-stealing means that there would be less need for `blocking` as long-lasting CPU compute operations could steal a lot of cycles and prevent lightweight I/O from being done even with `tokio-io-pool`. This thought just occurred to me so I thought I'd just post this here as an example of something that would benefit from work-stealing.
